standards/application/waku-api.md

```diff
@@ -0,0 +1,224 @@
+
+`Multiaddr` - a self-describing format for encoding network address of a remote peer as per [libp2p addressing](https://github.com/libp2p/specs/blob/6d38f88f7b2d16b0e4489298bcd0737a6d704f7e/addressing/README.md) specification.
+
+## API design
+
+### IDL
+
+A custom Interface Definition Language (IDL) in YAML is used to define the Waku API.
+Existing IDL Such as OpenAPI, AsyncAPI or WIT do not exactly fit the requirements for this API.
+Hence, instead of having the reader learn a new IDL, we propose to use a simple IDL with self-describing syntax.
+
+An alternative would be to choose a programming language. However, such choice may express unintended opinions on the API.
+
+#### Guidelines
+
+- No `default` means that the value is mandatory
+- Primitive types are `string`, `int`, `bool`, `enum` and `uint`
+- Complex pre-defined types are:
+  - `struct`: object and other nested types
+  - `option`: a value that can be set or left null
+  - `array`: iterable object containing values of all the same type
+  - `multiaddr`: a libp2p multiaddress; may be an object or a string, most idiomatic approach depending on the language
+  - `result`: an enum type that either contain a return value (success), or an error (failure). The error is left to the implementor
+  - `error`: Left to the implementor on whether `error` types are `string` or `object` in the given language.
+- The first parameter of a function MAY be considered as the object on which a method is called. It is left to the implementor on whether an objective programming approach is idiomatic to the language.
+
+### Application
+
+This API is designed for generic use and ease across all programming languages and traditional `REST` architectures.
+
+## The Waku API
+
+```yaml
+api_version: "0.0.1"
+library_name: "waku"
+description: "Waku: a private and censorship-resistant message routing library."
+```
+### Language Mappings
+
+How the API definition should be translated to specific languages.
+
+```yaml
+language_mappings:
+  rust:
+    naming_convention:
+      - functions: "snake_case"
+      - variables: "snake_case"
+      - types: "PascalCase"
+    
+  golang: 
+    naming_convention:
+       - functions: "snake_case"
+       - variables: "snake_case"
+       - types: "PascalCase"
+    
+  c:
+    naming_convention: "snake_case"
+    prefix: "waku_"
+    
+  typescript:
+    naming_convention:
+      - functions: "camelCase"
+      - variables: "camelCase"
+      - types: "PascalCase"
+    module_type: "esm"
+```
+
+
+### Type definitions
+
+```yaml
+types:
+  WakuNode:
+    type: struct
+    description: "A Waku node instance."
+  
+  Config:
+    type: struct
+    fields:
+      mode: 
+        type: string
+        constraints: ["edge", "relay"]
+        description: "The mode of operation of the Waku node. Core protocols used by the node are inferred from this mode."
+      network_config:
```

Comment:
Should this be "ClusterConf" instead? as network conf may related to TCP/IP level conf?